### PR TITLE
TLS support

### PR DESCRIPTION
**Pull request by KangJH**
## Main changes
- Support TLS handshake.
- Add Dev CA
- relocate sensr_proto. (latest one)
[Jira](https://seoulrobotics.atlassian.net/browse/AR-2848)

### CODE REVIEW

File: javascript/javascript_sdk/src/sensr_client.js [hunk 1]
```diff
@@ -1,11 +1,21 @@
 const WebSocket = require('ws');
 const outputMsg = require('./../js_proto/sensr_proto/output_pb.js');
 const pointMsg = require('./../js_proto/sensr_proto/point_cloud_pb');
+const fs = require('fs');
+
+const cert_location = __dirname + '/../../../' + 'keys/sensr-sdk-ca.crt'
```
Comment (thor-sr) [inline javascript/javascript_sdk/src/sensr_client.js:6]: '/../../../' :feelsbadman: but I can't think of any cleaner way of doing this.

File: javascript/javascript_sdk/src/sensr_client.js [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
+
+process.env.NODE_TLS_REJECT_UNAUTHORIZED='0'
```
Comment (thor-sr) [inline javascript/javascript_sdk/src/sensr_client.js:8]: What is this?

File: python/sensr_message_listener.py [hunk 3]
```diff
@@ -19,11 +21,15 @@ def __init__(self,
                  address="localhost", 
                  listener_type=ListenerType.BOTH, 
                  output_port = "5050", 
-                 point_port = "5051"):
-        self._address = "ws://" + address
+                 point_port = "5051",
+                 crt_file_path=""):
```
Comment (thor-sr) [inline python/sensr_message_listener.py:25]: Why is this not `"../keys/sensr-sdk-ca.crt"` by default?

File: src/websocket_endpoint.cpp [hunk 4]
```diff
@@ -76,6 +82,30 @@ namespace sensr
...
+          this,
+          std::placeholders::_1, 
+          std::placeholders::_2));
+
+        // Here we load the CA certificates of all CA's that this client trusts.
+        ctx->load_verify_file(cert_path_);
+    } catch (std::exception& e) {
+        std::cout << e.what() << std::endl;
+    }
+    return ctx;
```
Comment (thor-sr) [inline src/websocket_endpoint.cpp:106]: Is it fine to return this if the exception is raised?  Maybe we should change it to `nullptr`
```suggestion
    } catch (std::exception& e) {
        std::cout << e.what() << std::endl;
        ctx = nullptr;
    }
    return ctx;
```

File: src/websocket_endpoint.cpp [hunk 5]
```diff
@@ -108,5 +138,108 @@ namespace sensr
     }     
   }
 
+  /// Verify that one of the subject alternative names matches the given hostname
+  bool WebSocketEndPoint::VerifySubjectAlternativeName(const char * hostname, X509 * cert) {
+    STACK_OF(GENERAL_NAME) * san_names = NULL;
```
Comment (thor-sr) [inline src/websocket_endpoint.cpp:143]: I assume `STACK_OF(GENERAL_NAME)` is a macro that provides the typename, right? (Also let's change NULL to `nullptr`

File: src/websocket_endpoint.cpp [hunk 6]
```diff
@@ -108,5 +138,108 @@ namespace sensr
     }     
   }
 
+  /// Verify that one of the subject alternative names matches the given hostname
+  bool WebSocketEndPoint::VerifySubjectAlternativeName(const char * hostname, X509 * cert) {
+    STACK_OF(GENERAL_NAME) * san_names = NULL;    
+    san_names = (STACK_OF(GENERAL_NAME) *) X509_get_ext_d2i(cert, NID_subject_alt_name, NULL, NULL);
```
Comment (thor-sr) [inline src/websocket_endpoint.cpp:144]: If I'm understanding this weird code correctly, this should work
```suggestion
    san_names = static_cast<STACK_OF(GENERAL_NAME) *>(X509_get_ext_d2i(cert, NID_subject_alt_name, nullptr, nullptr));
```

File: src/websocket_endpoint.cpp [hunk 7]
```diff
@@ -108,5 +138,108 @@ namespace sensr
...
+    if (san_names == NULL) {
```
Comment (thor-sr) [inline src/websocket_endpoint.cpp:145]:
```suggestion
    if (san_names == nullptr) {
```

File: src/websocket_endpoint.cpp [hunk 8]
```diff
@@ -108,5 +138,108 @@ namespace sensr
...
+        return false;
+    }    
+    int san_names_count = sk_GENERAL_NAME_num(san_names);    
+    bool result = false;    
+    for (int i = 0; i < san_names_count; i++) {
+        const GENERAL_NAME * current_name = sk_GENERAL_NAME_value(san_names, i);        
+        if (current_name->type != GEN_DNS) {
+            continue;
+        }        
+        char const * dns_name = (char const *) ASN1_STRING_get0_data(current_name->d.dNSName);
```
Comment (thor-sr) [inline src/websocket_endpoint.cpp:155]: Can we static_cast?

File: src/websocket_endpoint.cpp [hunk 9]
```diff
@@ -108,5 +138,108 @@ namespace sensr
...
+    X509_NAME_ENTRY * common_name_entry = X509_NAME_get_entry(X509_get_subject_name(cert), common_name_loc);
+    if (common_name_entry == NULL) {
+        return false;
+    }    
+    // Convert the CN field to a C string
+    ASN1_STRING * common_name_asn1 = X509_NAME_ENTRY_get_data(common_name_entry);
+    if (common_name_asn1 == NULL) {
+        return false;
+    }    
+    char const * common_name_str = (char const *) ASN1_STRING_get0_data(common_name_asn1);
```
Comment (thor-sr) [inline src/websocket_endpoint.cpp:184]: Can we use static_cast?

**Review by thor-sr — APPROVED** [review]
Just a few comments, otherwise it looks great

File: javascript/javascript_sdk/src/sensr_client.js (hunk 1 above)
Comment (KangJH) [reply javascript/javascript_sdk/src/sensr_client.js:6]: Yes, sorry about this ugly style.

File: javascript/javascript_sdk/src/sensr_client.js (hunk 2 above)
Comment (KangJH) [reply javascript/javascript_sdk/src/sensr_client.js:8]: @WJ-sr Could you explain this?

File: python/sensr_message_listener.py (hunk 3 above)
Comment (KangJH) [reply python/sensr_message_listener.py:25]: At the first time I tried to set default path. But I faced a problem which the path depends on where the script was called.
So, I removed the default path and wanted to user to set the path correctly in their script.

File: src/websocket_endpoint.cpp (hunk 4 above)
Comment (KangJH) [reply src/websocket_endpoint.cpp:106]: let me try to do it.

File: src/websocket_endpoint.cpp (hunk 9 above)
Comment (KangJH) [reply src/websocket_endpoint.cpp:184]: sorry compiler complain invalid type conversion.
I think we need to use reinterpret_cast

File: src/websocket_endpoint.cpp (hunk 4 above)
Comment (KangJH) [reply src/websocket_endpoint.cpp:106]: I think this change made the fix of https://seoulrobotics.atlassian.net/browse/AR-2853

**Pull request merged**